File: pkg/payload/outgoing/attachment.go
package outgoing

import (
	"time"
)

type (
	Attachment struct {
		ID         string      `json:"attachmentID"`
		UserID     string      `json:"userID"`
		Url        string      `json:"url"`
		PreviewUrl string      `json:"previewUrl,omitempty"`
		Meta       interface{} `json:"meta"`
		Name       string      `json:"name"`
		CreatedAt  time.Time   `json:"createdAt,omitempty"`
		UpdatedAt  *time.Time  `json:"updatedAt,omitempty"`
	}

	AttachmentSet []*Attachment
)

File: lib/js/src/shared/types/attachment.ts
import { CortezaID, NoID, ISO8601Date, Apply } from '../../cast'
import { IsOf } from '../../guards'

interface Meta {
  [key: string]: unknown;
}

interface PartialAttachment extends Partial<Omit<Attachment, 'createdAt' | 'updatedAt' | 'deletedAt'>> {
  createdAt?: string|number|Date;
  updatedAt?: string|number|Date;
  deletedAt?: string|number|Date;
}

export class Attachment {
  public attachmentID = NoID;
  public ownerID = NoID;
  public name = '';
  public url = '';
  public previewUrl = '';
  public download = '';
  public meta: Meta = {};

  public createdAt?: Date = undefined;
  public updatedAt?: Date = undefined;
  public deletedAt?: Date = undefined;

  constructor (i?: PartialAttachment, baseURL?: string) {
    this.apply(i)
    this.setBaseURL(baseURL || '')
  }

  apply (i?: PartialAttachment): void {
    Apply(this, i, CortezaID, 'attachmentID', 'ownerID')
    Apply(this, i, String, 'name', 'url', 'previewUrl')

    if (IsOf(i, 'meta')) {
      this.meta = { ...i.meta }
    }

    Apply(this, i, ISO8601Date, 'createdAt', 'updatedAt', 'deletedAt')
  }

  setBaseURL (baseURL: string): void {
    this.url = baseURL + this.url
    this.previewUrl = baseURL + this.previewUrl
    this.download = this.url + '&download=1'
  }
}
